### PR TITLE
feat: add data loading

### PR DESCRIPTION
**Pull request by Bartosz7**

### CODE REVIEW

File: src/datasets/training.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import lightning.pytorch as pl
+from lightning.pytorch.loggers import WandbLogger
+import torch
+import os
+import sys
+from torchvision import transforms
+from torch.utils.data import DataLoader, ImageFolder
+from .utils import download_data
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
```
Comment (Simonexc) [inline src/datasets/training.py]: You don't need this line as long as this file is imported and run from the `src` directory. Since we should never just run this file, we don't need this line

File: src/datasets/training.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        if sys.platform in ["linux", "darwin"]:
+            data["num_workers"] = min(
+                len(os.sched_getaffinity(0)), 8
+            )  # num of cpu cores
+        return data
+
+    def prepare_data(self) -> None:
+        download_data(self.logger)
+        transform = transforms.Compose(
+            [transforms.Resize(128, 128), transforms.ToTensor()]
```
Comment (Simonexc) [inline src/datasets/training.py]: Let's make this customizable by `__init__` argument.

There are more things that we might want to do here. We should normalize the image to ensure that values are between -1 and 1. We first need to calculate the mean and std for all colors on the whole dataset. The mean and std should be saved in `project_config`. 

[Here](https://huggingface.co/docs/diffusers/tutorials/basic_training#load-the-dataset) they also include horizontal flip (since it should also provide a valid image). We can add a boolean flag for it.

File: src/datasets/training.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+from torch.utils.data import TensorDataset, DataLoader
+import lightning.pytorch as pl
+from lightning.pytorch.loggers import WandbLogger
+import torch
+import os
+import sys
+from torchvision import transforms
+from torch.utils.data import DataLoader, ImageFolder
```
Comment (Simonexc) [inline src/datasets/training.py]: I think `ImageFolder` should be imported from `torchvision.datasets`: https://pytorch.org/vision/stable/generated/torchvision.datasets.ImageFolder.html

Also, duplicated import of `DataLoader`

File: src/datasets/training.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+from project_config import config
+
+
+class TrainingDataset(pl.LightningDataModule):
+    def __init__(self, wandb_logger: WandbLogger, batch_size: int):
+        super().__init__()
+        self.logger = wandb_logger
+        self.batch_size = batch_size
+        self.train: TensorDataset | None = None
```
Comment (Simonexc) [inline src/datasets/training.py]: This is now `ImageFolder` class, or alternatively just `Dataset`

File: src/datasets/training.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            [transforms.Resize(128, 128), transforms.ToTensor()]
+        )
+        self.train = ImageFolder(config.cache_folder, transform=transform)
+
+    def train_dataloader(self) -> DataLoader:
+        return DataLoader(
+            self.train,
+            batch_size=self.batch_size,
+            **self.data_loader_kwargs,
+        )
```
Comment (Simonexc) [inline src/datasets/training.py]: We should add `shuffle=True`

File: src/datasets/training.py (hunk 3 above)
Comment (Simonexc) [inline src/datasets/training.py]: nit pick: The usual convention is to split imports into 3 groups (from top to bottom): standard library (os, sys); third-party imports (torch, etc.); application-specific imports (.utils, etc.). Each of these groups should be separated by one empty line. After all imports there should be two empty lines.

I mention it just for the readability. Probably I created some files which doesn't follow this logic, but it's just a good practice to follow which helps avoiding duplicate imports.

File: src/datasets/utils.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import sys
+import os
+from glob import glob
+from pathlib import Path
+import zipfile
```
Comment (Simonexc) [inline src/datasets/utils.py:2]: Is this in standard library?

File: src/datasets/utils.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    ):
+        try:
+            shutil.move(image_path, dummy_class_folder)
+        except shutil.Error:
+            print(f"Error moving {image_path}. Duplicate file.")
+    # remove the empty `data` folder
+    shutil.rmtree(os.path.join(config.cache_folder, "data"), ignore_errors=True)
+
+
+def download_data(wandb_logger: WandbLogger) -> tuple[np.ndarray, np.ndarray]:
```
Comment (Simonexc) [inline src/datasets/utils.py]: It looks like it doesn't return anything

File: src/datasets/utils.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        project=config.project, entity=config.entity, job_type=JobType.UPLOAD_DATA.value
+    ) as run:
+        artifact = wandb.Artifact(
+            config.dataset_artifact_name,
+            type=ArtifactType.DATASET.value,
+            description="Raw LSUN bedroom dataset",
+        )
+        # Load image files
+        image_paths = list(Path(config.data_folder).rglob("*.jpg"))
+        image_paths = image_paths[:100000]  # TODO: remove this line
```
Comment (Simonexc) [inline src/datasets/utils.py]: Remember to remove it and upload the whole dataset.

File: src/datasets/utils.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        # Add the zip file to the artifact
+        artifact.add_file(zip_filename)
+
+        # Upload to W&B
+        print("Uploading artifact...", end="")
+        run.log_artifact(artifact)
+        print("DONE")
+
+        # Clean up
+        os.remove(zip_filename)
```
Comment (Simonexc) [inline src/datasets/utils.py]: There are two issues with this. First, it creates zip file in the same directory as the code. I think we should just create it in `data` folder. 

The second problem is when something fails after the creation of zip. In that scenario the zip would not be removed. It's not a big issue but there is a simple fix for that. We could move it to a separate function which is a context manager ([example](https://book.pythontips.com/en/latest/context_managers.html#implementing-a-context-manager-as-a-generator)). This way we can easily ensure whatever happens, the zip will be removed. Alternatively, we can just put artifact upload code inside `try` and `os.remove` inside `finally` block.

File: src/datasets/utils.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    # Check if new data was downloaded
+    if artifacts_before != artifacts_after:
+        # Get the .zip file from the data_dir
+        zip_path = glob(os.path.join(data_dir, "*.zip"))[0]
+        print("Overwriting cache with new data...", end="")
+        unzip_to_cache(zip_path)
+        print("DONE")
+
+
+def upload_data():
```
Comment (Simonexc) [inline src/datasets/utils.py]: We want to execute `upload_data` as script. However, as a general rule we should only trigger scripts from the root source (in our case `src`) directory. We can either move this code to a script `src/upload_data.py` or just call this function there inside `if __name__ == "__main__"`.
This will solve issue with imports

File: src/datasets/utils.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        os.remove(zip_filename)
+
+
+if __name__ == "__main__":
+    pass
+    # for testing upload
+    # upload_data()
+    # for testing download
+    # with wandb.init(project=config.project, entity=config.entity) as run:
+    #     download_data(run)
```
Comment (Simonexc) [inline src/datasets/utils.py]: In general, we should never directly call files that are in subfolders.

File: src/trainer/prepare_session.py [hunk 12]
```diff
@@ -4,20 +4,21 @@
...
 from configs.run_config_class import RunConfig
+from dataset.training import TrainingDataset
 
 
 def prepare_session(
     run_config: RunConfig,
     wandb_logger: WandbLogger,
 ) -> tuple[pl.Trainer, None, None]:  # TODO: model and dataset to be implemented
-    data = None  # TODO: add loading data here
+    data = TrainingDataset(wandb_logger, run_config.batch_size)
```
Comment (Simonexc) [inline src/trainer/prepare_session.py]: Please add `batch_size` to `RunConfig` class

**Review by Simonexc — CHANGES_REQUESTED** [review]
If linter in your IDE doesn't correctly recognize imports, try to mark `src` folder as the source directory (it is definitely possible in PyCharm, I think in VSCode also).

File: src/datasets/training.py (hunk 3 above)
Comment (Bartosz7) [reply src/datasets/training.py]: Yep, I normally use linters to catch such quick mistakes, i was in a hurry then

File: src/datasets/utils.py (hunk 6 above)
Comment (Bartosz7) [reply src/datasets/utils.py:2]: yes, what's even more surprising is that it's  (`zipfile`) been built-in for quite some time already (Python 3.8+)

File: src/datasets/utils.py (hunk 7 above)
Comment (Bartosz7) [reply src/datasets/utils.py]: True, didn't check the type annotations, will improve

File: src/datasets/utils.py (hunk 8 above)
Comment (Bartosz7) [reply src/datasets/utils.py]: Yes, I will, but you may need to run the data upload yourself, as i have very limited space on my ubuntu partition

File: src/datasets/utils.py (hunk 10 above)
Comment (Bartosz7) [reply src/datasets/utils.py]: Will move

File: src/datasets/utils.py (hunk 8 above)
Comment (Simonexc) [reply src/datasets/utils.py]: Ok sure, will do

**Review by Simonexc — APPROVED** [review]
